plugin/pkg/admission/eventratelimit/admission.go

```diff
@@ -0,0 +1,270 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventratelimit
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	"github.com/golang/groupcache/lru"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/kubernetes/pkg/api"
+	eventratelimitapi "k8s.io/kubernetes/plugin/pkg/admission/eventratelimit/apis/eventratelimit"
+	"k8s.io/kubernetes/plugin/pkg/admission/eventratelimit/apis/eventratelimit/validation"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register("EventRateLimit",
+		func(config io.Reader) (admission.Interface, error) {
+			// load the configuration provided (if any)
+			configuration, err := LoadConfiguration(config)
+			if err != nil {
+				return nil, err
+			}
+			// validate the configuration (if any)
+			if configuration != nil {
+				if errs := validation.ValidateConfiguration(configuration); len(errs) != 0 {
+					return nil, errs.ToAggregate()
+				}
+			}
+			return newEventRateLimit(configuration)
+		})
+}
+
+// eventRateLimitAdmission implements an admission controller that can enforce event rate limits
+type eventRateLimitAdmission struct {
+	sync.RWMutex
+	*admission.Handler
+	limitEnforcers []*limitEnforcer
+}
+
+// limitEnforcer enforces a single type of event rate limit, such as server, namespace, or source+object
+type limitEnforcer struct {
+	// type of this limit
+	limitType string
+	// factory for creating a rate limiter
+	newRateLimiter func() flowcontrol.RateLimiter
+	// cache for holding the rate limiters
+	cache cache
+	// function for selecting the cache key with which an event is associated
+	keySelector func(admission.Attributes) interface{}
+}
+
+func newLimitEnforcer(config eventratelimitapi.Limit, rlf rateLimiterFactory) (*limitEnforcer, error) {
+	limitType := config.Type
+	newRateLimiter := func() flowcontrol.RateLimiter {
+		return rlf(config.QPS, config.Burst)
```

Comment:
This doesn't look valuable.  Why not just pass the `func() flowcontrol.RateLimiter` you need?

Reply:
I am not quite sure what you are suggesting here. I may have addressed, or avoided, this with other changes. But you may want to check that this specifically is addressed in the next revision of changes.